### ISSUE TITLE
 当graphic支持缩放的同时动态更新数据会发生异常

### ISSUE THREAD

**Issue by Ruifen9** [entronad/graphic#52]
Unhandled Exception: type 'ChangeDataSignal<ChartData>' is not a subtype of type 'GestureSignal' in type cast

**Comment by Ruifen9** [comment]
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: type 'ChangeDataSignal<ChartData>' is not a subtype of type 'GestureSignal' in type cast
E/flutter (24412): #0      _getRangeUpdate.<anonymous closure> (package:graphic/src/common/defaults.dart:19:14)
E/flutter (24412): #1      SignalUpdateOp.evaluate (package:graphic/src/interaction/signal.dart:93:18)
E/flutter (24412): #2      Operator.run (package:graphic/src/dataflow/operator.dart:95:29)
E/flutter (24412): #3      Dataflow.evaluate (package:graphic/src/dataflow/dataflow.dart:94:27)
E/flutter (24412): #4      View.evaluate (package:graphic/src/chart/view.dart:78:17)
E/flutter (24412): #5      Dataflow.run (package:graphic/src/dataflow/dataflow.dart:122:16)
E/flutter (24412): #6      Dataflow.bindChannel.<anonymous closure> (package:graphic/src/dataflow/dataflow.dart:154:7)
E/flutter (24412): #7      _rootRunUnary (dart:async/zone.dart:1436:47)
E/flutter (24412): #8      _CustomZone.runUnary (dart:async/zone.dart:1335:19)
E/flutter (24412): #9      _CustomZone.runUnaryGuarded (dart:async/zone.dart:1244:7)
E/flutter (24412): #10     _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:341:11)
E/flutter (24412): #11     _DelayedData.perform (dart:async/stream_impl.dart:591:14)
E/flutter (24412): #12     _StreamImplEvents.handleNext (dart:async/stream_impl.dart:706:11)
E/flutter (24412): #13     _PendingEvents.schedule.<anonymous closure> (dart:async/stream_impl.dart:663:7)
E/flutter (24412): #14     _rootRun (dart:async/zone.dart:1420:47)
E/flutter (24412): #15     _CustomZone.run (dart:async/zone.dart:1328:19)
E/flutter (24412): #16     _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
E/flutter (24412): #17     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1276:23)
E/flutter (24412): #18     _rootRun (dart:async/zone.dart:1428:13)
E/flutter (24412): #19     _CustomZone.run (dart:async/zone.dart:1328:19)
E/flutter (24412): #20     _CustomZone.runGuarded (dart:async/zone.dart:1236:7)
E/flutter (24412): #21     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1276:23)
E/flutter (24412): #22     _microtaskLoop (dart:async/schedule_microtask.dart:40:21)
E/flutter (24412): #23     _startMicrotaskLoop (dart:async/schedule_microtask.dart:49:5)

**Comment by Ruifen9** [comment]
``` dart
import 'package:flutter/material.dart';
import 'package:graphic/graphic.dart';
import 'package:watch_app/page/test/web_echart.dart';

class GraphicLineChart extends StatefulWidget{

  List<ChartData> items;


  GraphicLineChart(this.items);

  @override
  State<StatefulWidget> createState() {
    return _LineChartState();
  }

}

class _LineChartState extends State<GraphicLineChart>{

  @override
  Widget build(BuildContext context) {
    return Container(
      margin: const EdgeInsets.only(top: 10),
      width: 350,
      height: 150,
      child: Chart(
        padding: (_) => const EdgeInsets.fromLTRB(40, 5, 10, 0),
        rebuild: false,
        data: widget.items,
        variables: {
          'time': Variable(
            accessor: (ChartData data) => data.x,
            scale: TimeScale(tickCount: 5),
          ),
          'end': Variable(
            accessor: (ChartData data) => data.y,
            scale: LinearScale(min: 5, tickCount: 5),
          ),
        },
        elements: [
          LineElement(
            size: SizeAttr(value: 1),
          )
        ],
        axes: [
          Defaults.horizontalAxis
            ..label = null
            ..line = null,
          Defaults.verticalAxis
            ..gridMapper = (_, index, __) =>
            index == 0 ? null : Defaults.strokeStyle,
        ],
        selections: {
          'touchMove': PointSelection(
            on: {
              GestureType.scaleUpdate,
              GestureType.tapDown,
              GestureType.longPressMoveUpdate
            },
            dim: Dim.x,
          )
        },
///添加horizontalRangeUpdater: Defaults.horizontalRangeSignal后，更新数据会发生异常
        coord: RectCoord(
          horizontalRangeUpdater: Defaults.horizontalRangeSignal,
        ),
        crosshair: CrosshairGuide(
          followPointer: [true, false],
          styles: [
            StrokeStyle(color: const Color(0xffbfbfbf), dash: [4, 2]),
            StrokeStyle(color: const Color(0xffbfbfbf), dash: [4, 2]),
          ],
        ),
      ),
    );
  }

}

class ChartData {
  DateTime x;
  double y;

  ChartData(this.x, this.y);
}
```

**Comment by anggaaryas** [comment]
my solution is:
1. make rebuild: true, or
2. use try catch in horizontal range updater

**Comment by Ruifen9** [comment]
> my solution is:
> 
> 1. make rebuild: true, or
> 2. use try catch in horizontal range updater

Thank you for your reply.

1. make rebuild: true，Zoom will be reset
2. This is a temporary solution

**Comment by anggaaryas** [comment]
1. save the coord / pre and then assign to `horizontalRange`

**Comment by entronad** [comment]
It has been fixed in v0.9.1
